### PR TITLE
Phase change patch 4: Adding cell centered mass integrator using RK1/2 schemes to find the convective flux related to a transport quantity

### PR DESCRIPTION
**Pull request by amneetb**
### IBAMR Pull Request Checklist
- [ ] Does the test suite pass?
- [ ] Was clang-format run?
- [ ] Were relevant issues cited? Please remember to add `Fixes #12345` to close
      the issue automatically if we are fixing the problem.
- [ ] Is this a change others will want to know about? If so, then has a
      changelog entry been added?
- [ ] If new data structures have been added to a class then have they been set
      up appropriately for restarts? If so, ensure that the restart version
      number is incremented.
- [ ] Does this change include a bug fix or new functionality? If so, a new test
      or tests should be added. New tests should run quickly (less than a minute
      in release mode). If possible, an older test should gain a new option so
      that we do not need to compile more test executables.
- [ ] Did you (if your account has permission to do so) set relevant labels on
      GitHub for the pull request?

### CODE REVIEW

**Comment by Raamkrishnan** [comment]
Rebased this with the master and made the necessary changes in the ```AdvDiffConservativeMassScalarTransportRKIntegrator``` class. Please take a look.

File: include/ibamr/AdvDiffConservativeMassScalarTransportRKIntegrator.h [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+//
+// ---------------------------------------------------------------------
+
+/////////////////////////////// INCLUDE GUARD ////////////////////////////////
+
+#ifndef included_IBAMR_AdvDiffConservativeMassScalarTransportRKIntegrator
+#define included_IBAMR_AdvDiffConservativeMassScalarTransportRKIntegrator
+
+/////////////////////////////// INCLUDES /////////////////////////////////////
+#include "ibamr/STSMassFluxIntegrator.h"
```
Comment (drwells) [inline include/ibamr/AdvDiffConservativeMassScalarTransportRKIntegrator.h:20]:
```suggestion
#include <ibamr/config.h>

#include "ibamr/STSMassFluxIntegrator.h"
```
I need to finish the CI check to verify that the config header is always present - we always need it.

File: include/ibamr/AdvDiffConservativeMassScalarTransportRKIntegrator.h [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+ *
+ * and computes the conservative form of the convective operator
+ * \f$ \nabla \cdot (\rho \gamma u Q)\f$ where \f$\gamma\f$ is a material property
+ * and \f$Q\f$ is a transport quantity that is being transported. If \f$\gamma\f$ is not
+ * registered to this class, then this class computes \f$ \nabla \cdot (\rho u Q)\f$.
+ *
+ * This class implements the Forward Euler (RK-1) for single cycle and midpoint rule (RK-2) for
+ * multiple cycles as the time-stepping schemes.
+ *
+ * Class AdvDiffConservativeMassScalarTransportRKIntegrator computes the convective
```
Comment (drwells) [inline include/ibamr/AdvDiffConservativeMassScalarTransportRKIntegrator.h:56]:
```suggestion
 * This class computes the convective
```

File: include/ibamr/AdvDiffConservativeMassScalarTransportRKIntegrator.h [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+        const double& dt,
+        const double* const dx);
+
+    // Book keeping
+    std::string d_object_name;
+
+    // Cached communications operators.
+    std::string d_transport_quantity_bdry_extrap_type = "CONSTANT", d_material_property_bdry_extrap_type = "CONSTANT";
+
+    std::vector<IBTK::HierarchyGhostCellInterpolation::InterpolationTransactionComponent> d_Q_transaction_comps,
```
Comment (drwells) [inline include/ibamr/AdvDiffConservativeMassScalarTransportRKIntegrator.h:262]:
```suggestion
    std::string d_transport_quantity_bdry_extrap_type = "CONSTANT", d_material_property_bdry_extrap_type = "CONSTANT";

    // Cached communications operators.
    std::vector<IBTK::HierarchyGhostCellInterpolation::InterpolationTransactionComponent> d_Q_transaction_comps,
```

**Review by drwells — CHANGES_REQUESTED** [review]
It looks like there aren't any tests for this class (though it is used indirectly in the tests in #1628). Could you add one which just integrates some quantity and then prints it out on a coarse grid?

**Comment by Raamkrishnan** [comment]
> It looks like there aren't any tests for this class (though it is used indirectly in the tests in #1628). Could you add one which just integrates some quantity and then prints it out on a coarse grid?

This class is not a standalone integrator. It is tightly coupled to ```PhaseChangeHierarchyIntegrator``` class. I think the existing interface needs to be changed to add a test just for this class. As you noted, we have already added a test that uses both ```PhaseChangeHierarchyIntegrator``` and ```AdvDiffConservativeMassScalarTransportRKIntegrator.h``` in #1628.

**Comment by Raamkrishnan** [comment]
@drwells Ping

**Comment by amneetb** [comment]
@drwells : if all looks good can we merge this and move to the next patch?

**Comment by drwells** [comment]
I have some ideas for how to test this - I'll open an issue for that and lets keep going.

**Review by drwells — APPROVED** [review]

**Pull request merged**